File: 백준/Gold/3079. 입국심사/입국심사.py
import sys
input = sys.stdin.readline

# upper bound
def binary_search(left, right, n, target, times):    
    ans = right
    while(left <= right):
        mid = (left + right) // 2
        people = 0
    
        for i in range(n):
            people += mid // times[i]
        
        if people >= target: # 현재 처리할 수 있는 사람의 수 >= target -> 시간을 줄여보자
            right = mid - 1
        else:
            left = mid + 1
            
    return left

n, m = map(int, input().split())
# m명의 사람이 심사를 받는 데 걸리는 시간의 최소 구하기
times = list()
for _ in range(n):
    times.append(int(input()))
    
times.sort()
  
ans = binary_search(times[0], times[n-1] * m, n, m, times)    
print(ans)

File: 백준/Gold/3079. 입국심사/입국심사.cc
#include <iostream>
#include <vector>
#include <algorithm>

using namespace std;

typedef long long ll;

ll cntPeople(ll n, ll mid, vector<ll> &time){
    // mid 시간에 검사할 수 있는 최대 인원 수
    ll cnt = 0;
    for (int i = 0; i < n; i++) {
        cnt += mid / time[i];
    }
    return cnt;
}

ll binarySearch(ll n, ll m, vector<ll> &time) {
    ll left = 1;
    ll right = m * time[0];  // 심사 시간이 제일 짧은 데스크에 모든 사람이 검사 받는 경우를 최댓값으로 초기화
    ll mid;
    ll people;

    while (left <= right) {
        mid = (left + right) / 2;
        people = cntPeople(n, mid, time);

        // mid에서 검사 가능 인원이 m 명 이상일 시 -> 조건 만족, 심사시간을 줄여보자
        if (people >= m) {
            right = mid - 1;
        } else {
            left = mid + 1;
        }
    }
    return left;
}

int main() {
    // 입력
    ll n, m;
    cin >> n >> m;
    vector<ll> time(n);
        for (int i = 0; i < n; i++) {
        cin >> time[i];
    }

    sort(time.begin(), time.end());
    // 연산 & 출력
    cout << binarySearch(n, m, time);

    return 0;
}